### PR TITLE
Add telemetry notification

### LINKED ISSUE #439: Implement Telemetry Notification
To log errors/warnings in to telemetry we need to implement [telemetry event](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#telemetry_event)

Related to  https://github.com/redhat-developer/vscode-yaml/issues/458

### PR DESCRIPTION
**Pull request by evidolob**
### What does this PR do?
Add telemetry notification.
It log errors happened during codecompetion computation and also log if user has `kubernetes` shema mapping.

### What issues does this PR fix or reference?
Resolve #439

### CODE REVIEW

File: src/languageservice/services/yamlCompletion.ts [hunk 1]
```diff
@@ -141,6 +146,7 @@ export class YAMLCompletion extends JSONCompletion {
       },
       error: (message: string) => {
         console.error(message);
+        this.telemetry.sendError('competion_error', { message });
```
Comment (fbricon) [inline src/languageservice/services/yamlCompletion.ts:149]: yaml.completion.error

Should send {"error": message} as properties payload

File: src/languageservice/utils/schemaUrls.ts [hunk 2]
```diff
@@ -1,13 +1,20 @@
...
 
-export function checkSchemaURI(workspaceFolders: WorkspaceFolder[], workspaceRoot: URI, uri: string): string {
+export function checkSchemaURI(
+  workspaceFolders: WorkspaceFolder[],
+  workspaceRoot: URI,
+  uri: string,
+  telemetry: Telemetry
+): string {
   if (uri.trim().toLowerCase() === 'kubernetes') {
+    telemetry.send({ name: 'schema_configured', properties: { kubernetes: true } });
```
Comment (fbricon) [inline src/languageservice/utils/schemaUrls.ts:17]: 'yaml.schema.configured'

File: src/languageserver/telemetry.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+export class Telemetry {
+  constructor(private readonly connection: Connection) {}
+
+  send(event: TelemetryEvent): void {
+    this.connection.telemetry.logEvent(event);
+  }
+
+  sendError(name: string, properties: unknown): void {
+    this.send({ name, type: 'track', properties });
```
Comment (fbricon) [inline src/languageserver/telemetry.ts:28]: { name, type: 'track', properties: properties }

File: src/languageserver/telemetry.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  sendTrack(name: string, properties: unknown): void {
+    this.send({ name, type: 'track', properties });
```
Comment (fbricon) [inline src/languageserver/telemetry.ts:32]: { name, type: 'track', properties: properties }

File: src/languageserver/telemetry.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Red Hat, Inc. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Connection } from 'vscode-languageserver/lib/common/server';
+
+/**
+ * Doe to LSP limitation this object must be JSON serializable
```
Comment (JPinkney) [inline src/languageserver/telemetry.ts:9]:
```suggestion
 * Due to LSP limitation this object must be JSON serializable
```
minor nitpick

File: src/languageserver/telemetry.ts (hunk 3 above)
Comment (JPinkney) [inline src/languageserver/telemetry.ts:28]: Should this be:
```
this.send({ name, type: 'error', properties });
```
?
Currently it looks like this line and sendTrack are the same

Comment (fbricon) [reply src/languageserver/telemetry.ts:28]: no, this is how segment payloads must look like

Comment (fbricon) [reply src/languageserver/telemetry.ts:28]: now, we could argue wether the yaml.ls coupling to segment format is a good or a bad thing

File: src/languageserver/telemetry.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ *--------------------------------------------------------------------------------------------*/
+
+import { Connection } from 'vscode-languageserver/lib/common/server';
+
+/**
+ * Due to LSP limitation this object must be JSON serializable
+ */
+export interface TelemetryEvent {
+  type?: string;
+  name: string;
```
Comment (joshuawilson) [inline src/languageserver/telemetry.ts:13]: If the `name` is the only required field, shouldn't it go first?

**Comment by JPinkney** [comment]
Is it possible to add some tests for this to make sure errors are logged during code completion and kubernetes schema mapping?

**Review by joshuawilson — APPROVED** [review]

**Review by JPinkney — APPROVED** [review]

**Pull request merged**